#0: seperate dispatch constants

nhuang-tt opened the pull request:
### Ticket
#16339

### Problem description
`dispatch_constants` is actually a memory map.

### What's changed
- Rename `dispatch_constants` to `DispatchMemMap`
- Removed constants from `DispatchMemMap (ex dispatch_constants)`. They are already in DispatchConstants and it doesn't need the singleton.
- DispatchMemMap address sizes like Dispatch Buffer Size, Cmddat size, etc. are configured by DispatchSettings. Eventually DispatchSettings + FDTopology will be loaded from a file.

### Checklist
[Single Card](https://github.com/tenstorrent/tt-metal/actions/runs/13042437025)
[All Post Commit](https://github.com/tenstorrent/tt-metal/actions/runs/13042427528)
[T3K](https://github.com/tenstorrent/tt-metal/actions/runs/13042433714)
[TG/TGG](https://github.com/tenstorrent/tt-metal/actions/runs/13042431630)

nhuang-tt committed:
#0: seperate dispatch constants

- Rename dispatch_constants to DispatchMemMap
- Delete old constants. New constants in DispatchConstants
- DispatchMemMap to be configured by DispatchSettings

tt-dma commented on tt_metal/api/tt-metalium/command_queue.hpp:
```diff
@@ -172,15 +173,14 @@ class EnqueueTraceCommand : public Command {
+        std::array<uint32_t, DispatchConstants::DISPATCH_MESSAGE_ENTRIES>& expected_num_workers_completed,
         NOC noc_index,
         CoreCoord dispatch_core);
+       void process();
```
The indents here look a little off, or is it just me?

nhuang-tt replied:
Hmmmm looks like it got confused because `public:` and `private:` were not properly indented before. I fixed it now

nhuang-tt commented:
> So we have `DispatchMemMap`, `DispatchConstants`, and `DispatchSettings` now?
> 
> `DispatchMemMap` to represent addresses/sizes, derived from `DispatchConstants` and `DispatchSettings` `DispatchConstants` to hold static constants `DispatchSettings` to hold constants that can be set
> 
> Is that understanding correct?

Correct

Constants like `DISPATCH_MESSAGE_ENTRIES`, `DISPATCH_GO_SIGNAL_NOC_DATA_ENTRIES`, and especially the `TRANSFER_PAGE_SIZE` may change in the future depending on the system. 

It would be better to remove place those values into`DispatchSettings` but that's another issue to inject the settings object to where DispatchConstants is used.

tt-dma commented:
Do you think it makes sense to combine `DispatchConstants` and `DispatchSettings`? It might be simpler to just have `DispatchSettings` and some of them (all of them?) are settable

omilyutin-tt commented on tt_metal/api/tt-metalium/command_queue_interface.hpp:
```diff
@@ -43,64 +44,27 @@ enum class CommandQueueHostAddrType : uint8_t {
             hw_cqs = num_hw_cqs;
             last_core_type = core_type;
-            inst = std::unique_ptr<dispatch_constants>(new dispatch_constants(core_type, hw_cqs));
+            inst = std::unique_ptr<DispatchMemMap>(new DispatchMemMap(core_type, hw_cqs));
```
Do you re-create the map every time either core type or hw_cqs change? Is this only used in tests?

omilyutin-tt replied:
Very similar code as https://github.com/tenstorrent/tt-metal/pull/17320 @tt-asaigal

nhuang-tt replied:
It's similar to the PR you linked. This mem map is used in the command queue and creating the fast dispatch kernels.

It gets re-created if the num hw CQs and core type combo doesn't match the last one. In my testing and when I ran the UNET model I only saw it get created once. it's not a supported use case to switch #CQs at runtime?

omilyutin-tt commented on tt_metal/api/tt-metalium/command_queue_interface.hpp:
```diff
@@ -43,64 +44,27 @@ enum class CommandQueueHostAddrType : uint8_t {
+    static const DispatchMemMap& get(const CoreType& core_type, const uint32_t num_hw_cqs = 0) {
         if (num_hw_cqs > 0 && (num_hw_cqs != hw_cqs || core_type != last_core_type || !inst)) {
             hw_cqs = num_hw_cqs;
             last_core_type = core_type;
-            inst = std::unique_ptr<dispatch_constants>(new dispatch_constants(core_type, hw_cqs));
+            inst = std::unique_ptr<DispatchMemMap>(new DispatchMemMap(core_type, hw_cqs));
```
Instead of the member variables, let's have an instance here as a `static` function local variable. Also no need to use the unique ptr here - will add an extra memory hop.

nhuang-tt replied:
Done

omilyutin-tt commented on tt_metal/api/tt-metalium/command_queue_interface.hpp:
```diff
@@ -43,64 +44,27 @@ enum class CommandQueueHostAddrType : uint8_t {
-// Contains constants related to FD
 //
-// Deprecated note: for constant values, use tt::tt_metal::dispatch::DispatchConstants instead.
+// Dispatch Memory Map
```
Can you add a couple of more details on this?

nhuang-tt replied:
Done

omilyutin-tt commented on tt_metal/api/tt-metalium/command_queue_interface.hpp:
```diff
@@ -43,64 +44,27 @@ enum class CommandQueueHostAddrType : uint8_t {
 //
-//
-struct [[deprecated]] dispatch_constants {
+struct DispatchMemMap {
```
Use `class` at this point

omilyutin-tt commented on tt_metal/api/tt-metalium/dispatch_constants.hpp:
```diff
@@ -10,7 +10,7 @@
 #include "cq_commands.hpp"
 #include "dev_msgs.h"
 
-namespace tt::tt_metal::dispatch {
```
Just checking that this is intentional

nhuang-tt replied:
Other dispatch code seems to be using `tt::tt_metal` without the `dispatch`

nhuang-tt commented:
> Do you think it makes sense to combine `DispatchConstants` and `DispatchSettings`? It might be simpler to just have `DispatchSettings` and some of them (all of them?) are settable

I want all/most DispatchConstants values to be available in `DIspatchSettings` as a configuration option / runtime.

We can move the `static constexpr` values from `DispatchConstants` into `DispatchSettings`, but they won't be available to be configured because the settings depend on `num_hw_cqs` and `dispatch_core_type`.

In a lot of cases where `DispatchConstants` is used num CQs and core type aren't available. To get it you would need to find and the `IDevice` in a weird way.

nhuang-tt committed:
#0: merge DispatchConstants to DispatchSettings

- DispatchMemMap updated to be a class. Removed unique_ptr
  for static local variable
- Use tt_align.hpp
- Merged static constexpr values to DispatchSettings

nhuang-tt commented:
@tt-dma I merged DispatchConstants to DispatchSettings

nhuang-tt committed:
#0: update DispatchMemMap comment

tt-dma reviewed (approved):
Looks good, thanks for the cleanup!

nhuang-tt committed:
#0: resolve conflict
